src/main/java/io/cdap/plugin/http/source/common/pagination/page/JSONUtil.java

```diff
@@ -28,5 +34,151 @@ public static JsonObject toJsonObject(String text) {
+   * @return an object containing information about search results, success/failure.
+   */
+  public static JsonQueryResponse getJsonElementByPath(JsonObject jsonObject, String jsonPath) {
+    String stripped = StringUtils.strip(jsonPath.trim(), "/");
+    String[] pathParts;
+    if (!stripped.isEmpty()) {
+      pathParts = stripped.split("/");
+    } else {
+      pathParts = new String[0];
+    }
+
+    JsonElement currentElement = jsonObject;
+    for (int i = 0; i < pathParts.length; i++) {
+      String pathPart = pathParts[i];
+
+      if (currentElement.isJsonObject()) {
+        jsonObject = currentElement.getAsJsonObject();
+      }
+
+      if (!currentElement.isJsonObject() || jsonObject.get(pathPart) == null) {
+        return new JsonQueryResponse(
+          Arrays.copyOfRange(pathParts, 0, i),
+          Arrays.copyOfRange(pathParts, i, pathParts.length),
+          currentElement
+        );
+      }
+
+      currentElement = jsonObject.get(pathPart);
+    }
+    return new JsonQueryResponse(
+      Arrays.copyOfRange(pathParts, 0, pathParts.length),
+      new String[0],
+      currentElement
+    );
+  }
+
+  /**
+   * A class which contains information regarding results of searching an element in json by json path.
+   */
+  public static class JsonQueryResponse {
+    private final String[] retrivedPathParts;
+    private final String[] unretrivedPathParts;
+    private final String retrivedPath;
+    private final String unretrivedPath;
+    private final JsonElement result;
+
+    public JsonQueryResponse(String[] retrivedPathParts, String[] unretrivedPathParts, JsonElement result) {
+      this.retrivedPathParts = retrivedPathParts;
+      this.unretrivedPathParts = unretrivedPathParts;
+      this.retrivedPath = "/" + StringUtils.join(retrivedPathParts, '/');
+      this.unretrivedPath = "/" + StringUtils.join(unretrivedPathParts, '/');
+      this.result = result;
+    }
+
+    /**
+     * Assert if an element found is or correct type (e.g. primitive/object/array)
+     *
+     * @param expectedClass a class representing a type of json element.
+     */
+    public void assertClass(Class<? extends JsonElement> expectedClass) {
+      if (!expectedClass.isInstance(result)) {
+        throw new IllegalStateException(String.format(
+          "Element retrieved by path '%s' expected to be '%s', but found '%s'.\nResult json is: '%s'",
+          getRetrivedPath(), expectedClass.getSimpleName(),
+          result.getClass().getSimpleName(), result.toString()));
+      }
+    }
+
+    /**
+     * @return true if the json path was fully successfully retrieved till the last element.
+     */
+    public boolean isFullyRetrieved() {
+      return (unretrivedPathParts.length == 0);
+    }
+
+    /**
+     * Fails if json path was not fully successfully retrieved till the last element.
+     */
+    public void assertFullyRetrieved() {
+      if (!isFullyRetrieved()) {
+        throw new IllegalStateException(String.format(
```

Comment:
this isn't really an illegal state, IllegalArgument fits a little better

Reply:
done